### ISSUE TITLE
[CLIFF] Details of finetuning the CLIFF Annotator

### ISSUE THREAD

**Issue by haofanwang** [huawei-noah/noah-research#121]
As title. @zhihaolee 

To generate pseudo-GT for datasets in the wild with 2D keypoints, could you provide more details how to use the predicted SMPL params as regularization and 2D kpts as weak supervision? As you have shared your pre-trained model, so the step (a) & (b) are already done. Is it the same procedure as SPIN via iterative fitting on 2D joints?

**Comment by zhihaolee** [comment]
No, it's different from SPIN. There is no SMPLify inside. It's like EFT, but in a batch manner and with SMPL regularization.

**Comment by Rainfalllen** [comment]
Is there something wrong with my code? I have exploding gradients at the beginning of training (2d loss is NaN).
```
    def j2d_processing_full(self, kp, center, scale, r, f, img_w, img_h):
        nparts = kp.shape[0]
        for i in range(nparts):
            kp[i, 0:2] = transform(kp[i, 0:2] + 1, center, scale,
                                   [self.options.IMG_RES, self.options.IMG_RES], rot=r)
        kp[:, :-1] = kp[:, :-1] - self.options.IMG_RES * 0.5
        # flip the x coordinates
        if f:
            kp = flip_kp(kp)

        img_center = [img_w / 2.0, img_h / 2.0]
        delta = center - img_center
        kp[:, :-1] = kp[:, :-1] + delta[None, :]
        kp[:, 0] = 2. * kp[:, 0] / img_w
        kp[:, 1] = 2. * kp[:, 1] / img_h

        kp = kp.astype('float32')
        return kp
```
```
        joints3d = smpl_output['smpl_joints3d']

        bs = joints3d.shape[0]
        device = joints3d.device
        joints2d = perspective_projection(
            joints3d,
            rotation=torch.eye(3, device=device).unsqueeze(0).expand(bs, -1, -1),
            translation=pred_cam_full,
            focal_length=focal_length,
            camera_center=torch.zeros(bs, 2, `device=device)`
        )

        res = torch.stack([img_w / 2., img_h / 2.], dim=-1)
        joints2d = joints2d / res[:, None, :]

        smpl_output['smpl_joints2d'] = joints2d
```
